File: transport_booking_fast/static/src/js/transport_booking.js
/** @odoo-module **/

import { registry } from "@web/core/registry";
import { Component, useState, onWillStart } from "@odoo/owl";
import { rpc } from "@web/core/network/rpc";

// CLASS DEFINITION
export class TransportBooking extends Component {
    setup() {
        this.state = useState({
            routes: [],
            vehicles: [],
            selectedRoute: null,
            selectedVehicle: null,
            customerName: "",
            seats: 1,
            amount: 0.0,
        });
        console.log("state....",this.state);

        onWillStart(async () => {
            try {
                // Direct call to controller route
                const data = await rpc("/transport/data");
                this.state.routes = data.routes;
                this.state.vehicles = data.vehicles;
                console.log("data....",data);
            } catch (error) {
                console.error("Failed to load master data:", error);

            }
        });
    }
     //  CUSTOM METHOD CALLED BY BUTTON WHEN CLICK
    async confirmBooking() {
        const bookingData = {
            route_id: parseInt(this.state.selectedRoute),
            vehicle_id: parseInt(this.state.selectedVehicle),
            customer_name: this.state.customerName,
            seats: parseInt(this.state.seats),
            amount: parseFloat(this.state.amount),
        };
        console.log("booking data....",bookingData);

        if (!bookingData.route_id || !bookingData.vehicle_id || !bookingData.customer_name) {
            alert("Please fill in all fields.");
            return;
        }

        try {
            // Direct call to controller route
            const result = await rpc("/transport/booking/create", { data: bookingData });
            if (result.success) {
                this.state.selectedRoute = null;
                this.state.selectedVehicle = null;
                this.state.customerName = "";
                this.state.seats = 1;
                this.state.amount = 0.0;
                alert("Booking Confirmed!");
                console.log("result....",result);
            }
        } catch (error) {
            console.error("Booking failed:", error);
            alert("An error occurred during booking.");
        }
    }
}
//  TEMPLATE MAPPING
TransportBooking.template = "transport_booking_fast.TransportBooking";
//console.log("templete....",template);
// ACTION REGISTRATION
registry.category("actions").add("transport_booking_fast.TransportBooking", TransportBooking);
console.log("transport booking....",TransportBooking);

File: transport_booking_fast/models/transport_booking.py
from odoo import models, fields, api

class TransportBooking(models.Model):
    _name = 'transport.booking'
    _description = 'Transport Booking'

    route_id = fields.Many2one('transport.route', string='Route', required=True)
    vehicle_id = fields.Many2one('transport.vehicle', string='Vehicle', required=True)
    booking_date = fields.Datetime(string='Booking Date', default=fields.Datetime.now, required=True)
    customer_name = fields.Char(string='Customer Name', required=True)
    seats = fields.Integer(string='Seats', required=True)
    amount = fields.Float(string='Amount', required=True)

    @api.model
    def create_booking(self, data):
        """Creates a booking record from the provided dictionary `data`."""
        record = self.create({
            'route_id': int(data.get('route_id')),
            'vehicle_id': int(data.get('vehicle_id')),
            'customer_name': data.get('customer_name'),
            'seats': int(data.get('seats')),
            'amount': float(data.get('amount')),
        })
        return {'success': True, 'id': record.id}
